pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,349 @@
+	out, err := os.Create(stagedUpdatePath)
+	if err != nil {
+		return "", fmt.Errorf("could not create file at %s: %w", stagedUpdatePath, err)
+	}
+	defer out.Close()
+
+	// Request download from mirror
+	resp, err := ulm.mirrorClient.Get(ulm.mirrorUrl + fmt.Sprintf("/kolide/%s/%s/%s", binary, runtime.GOOS, targetFilename))
+	if err != nil {
+		return stagedUpdatePath, fmt.Errorf("could not make request to download target %s: %w", targetFilename, err)
+	}
+	defer resp.Body.Close()
+
+	// Wrap the download in a LimitReader so we read at most localMeta.Length bytes
+	stream := io.LimitReader(resp.Body, localTargetMetadata.Length)
+	var fileBuffer bytes.Buffer
+
+	// Read the target file, simultaneously writing it to our file buffer and generating its metadata
+	actualTargetMeta, err := util.GenerateTargetFileMeta(io.TeeReader(stream, io.Writer(&fileBuffer)), localTargetMetadata.HashAlgorithms()...)
+	if err != nil {
+		return stagedUpdatePath, fmt.Errorf("could not write downloaded target %s to file %s and compute its metadata: %w", targetFilename, stagedUpdatePath, err)
+	}
+
+	// Verify the actual download against the confirmed local metadata
+	if err := util.TargetFileMetaEqual(actualTargetMeta, localTargetMetadata); err != nil {
+		return stagedUpdatePath, fmt.Errorf("verification failed for target %s staged at %s: %w", targetFilename, stagedUpdatePath, err)
+	}
+
+	// Everything looks good: write the buffered download to disk
+	if _, err := io.Copy(out, &fileBuffer); err != nil {
+		return stagedUpdatePath, fmt.Errorf("could not write downloaded target %s to file %s: %w", targetFilename, stagedUpdatePath, err)
+	}
+
+	return stagedUpdatePath, nil
+}
+
+// moveVerifiedUpdate untars the update, moves it into the update library, and performs final checks
+// to make sure that it's a valid, working update.
+func (ulm *updateLibraryManager) moveVerifiedUpdate(binary autoupdatableBinary, targetFilename string, stagedUpdate string) error {
+	targetVersion := ulm.versionFromTarget(binary, targetFilename)
+	newUpdateDirectory := filepath.Join(ulm.updatesDirectory(binary), targetVersion)
+	if err := os.MkdirAll(newUpdateDirectory, 0755); err != nil {
+		return fmt.Errorf("could not create directory %s for new update: %w", newUpdateDirectory, err)
+	}
+
+	if err := fsutil.UntarBundle(filepath.Join(newUpdateDirectory, string(binary)), stagedUpdate); err != nil {
+		ulm.removeUpdate(binary, targetVersion)
+		return fmt.Errorf("could not untar update to %s: %w", newUpdateDirectory, err)
+	}
+
+	// Make sure that the binary is executable
+	if err := os.Chmod(executableLocation(newUpdateDirectory, binary), 0755); err != nil {
+		ulm.removeUpdate(binary, targetVersion)
+		return fmt.Errorf("could not set +x permissions on executable: %w", err)
+	}
+
+	// Validate the executable
+	if err := autoupdate.CheckExecutable(context.TODO(), executableLocation(newUpdateDirectory, binary), "--version"); err != nil {
```

Comment:
I don't love that these checks are happening in the library path and not a staged or random path. But I can live with it. (we check the executability before invoking anyhow)

Reply:
Will address in https://github.com/kolide/launcher/pull/1119

Reply:
Addressed in https://github.com/kolide/launcher/pull/1119/commits/51286ad8923a7cf0562f2d9eeb291580af702e6f